# LINK to Github Repo: (https://github.com/RyanWri/afeka-nlp-homeworks)
## repo is private, request access if needed

## Section 1 + 2
***please see requirements.txt***

##  Section 3 + 4
## loads data and Compute simple statistics on the data

In [1]:
from analysis import load_data, analyze_spam_csv

In [2]:
# File path to the CSV file
file_path = "data/spam.csv"

data = load_data(file_path)

# Analyze and print statistics
stats = analyze_spam_csv(data)
for key, value in stats.items():
    print(f"{key}: {value}")

Number of SMS messages: 5572
Number of spam messages: 747
Total word count: 86835
Average words per message: 15.584
5 most frequent words: [('to', 2248), ('i', 2239), ('you', 2128), ('a', 1442), ('the', 1330)]
Number of rare words: 5276


## Section 5 Perform tokenization

In [3]:
from tokenize_nltk import tokenize_nltk

[nltk_data] Downloading package punkt to /Users/ranwright/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /Users/ranwright/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


In [4]:
# Tokenize using NLTK
nltk_tokens = tokenize_nltk(data["Message"])
print(f"Total NLTK tokens: {len(nltk_tokens)}")

Total NLTK tokens: 104904


In [5]:
from tokenize_spacy import tokenize_spacy

In [6]:
# Tokenize using spaCy
spacy_tokens = tokenize_spacy(data["Message"])
print(f"Total spaCy tokens: {len(spacy_tokens)}")

Total spaCy tokens: 104367


In [7]:
from analysis import compare_tokenizations

In [8]:
compare_token_method_result = compare_tokenizations(nltk_tokens, spacy_tokens)
for key, value in compare_token_method_result.items():
    print(f"{key}: {value}")

NLTK unique tokens count: 689
spaCy unique tokens count: 743
Common tokens count: 10822
Total NLTK tokens: 104904
Total spaCy tokens: 104367


## Section 6 Perform lemmatization

In [9]:
from lemmatization_nltk import lemmatize_nltk

[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/ranwright/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     /Users/ranwright/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/ranwright/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     /Users/ranwright/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger_eng is already up-to-
[nltk_data]       date!


In [10]:
# Lemmatize using NLTK
nltk_lemmas = lemmatize_nltk(data["Message"])
print(f"Total NLTK lemmas: {len(nltk_lemmas)}")

Total NLTK lemmas: 104904


In [11]:
from lemmatization_spacy import lemmatize_spacy

In [12]:
# Lemmatize using spaCy
spacy_lemmas = lemmatize_spacy(data["Message"])
print(f"Total spaCy lemmas: {len(spacy_lemmas)}")

Total spaCy lemmas: 104367


In [13]:
from analysis import compare_lemmatizations

In [14]:
# Compare lemmatizations
lemma_diff = compare_lemmatizations(nltk_lemmas, spacy_lemmas)
print("\nComparison of Lemmatizations:")
for key, value in lemma_diff.items():
    print(f"{key}: {value}")


Comparison of Lemmatizations:
NLTK unique lemmas count: 2379
spaCy unique lemmas count: 1434
Common lemmas count: 8221
Total NLTK lemmas: 104904
Total spaCy lemmas: 104367


## Section 7 Perform stemming

In [15]:
from stem_nltk import stem_nltk

[nltk_data] Downloading package punkt to /Users/ranwright/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [16]:
# Stem using NLTK (Porter Stemmer)
nltk_stems = stem_nltk(data["Message"], method="porter")
print(f"Total NLTK (Porter) stems: {len(nltk_stems)}")

Total NLTK (Porter) stems: 104904


In [17]:
from stem_spacy import stem_spacy

In [18]:
# Stem using spaCy (approximation)
spacy_stems = stem_spacy(data["Message"])
print(f"Total spaCy stems: {len(spacy_stems)}")

Total spaCy stems: 104367


In [19]:
from analysis import compare_stemmings

In [20]:
# Compare stemmings
stem_diff = compare_stemmings(nltk_stems, spacy_stems)
print("\nComparison of Stemmings:")
for key, value in stem_diff.items():
    print(f"{key}: {value}")


Comparison of Stemmings:
NLTK unique stems count: 5870
spaCy unique stems count: 4777
Common stems count: 2197
Total NLTK stems: 104904
Total spaCy stems: 104367


## Section 8 Finding a Message That Reduces Stems but Keeps Lemmas Unchanged

This analysis identifies a message from the dataset where removing it would:
- Reduce the total number of unique **stemmed tokens**.
- Keep the total number of unique **lemmatized tokens** unchanged.

The result highlights cases where stemming is more aggressive and fragmented compared to lemmatization, which tends to normalize words consistently to their base forms.


In [21]:
from comparison import find_message_with_stem_lemma_disparity

In [22]:
# Example usage
result_message = find_message_with_stem_lemma_disparity(
    messages=data["Message"].tolist(), nltk_stems=nltk_stems, nltk_lemmas=nltk_lemmas
)

if result_message:
    print("Message causing the effect:\n")
    print(
        repr(result_message)
    )  # Ensures the entire message, including multiline, is printed
else:
    print("No message found that meets the criteria.")

Message causing the effect:

'Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...'


## Section 9 Finding a Message That Reduces Lemmas but Keeps Stems Unchanged

This analysis identifies a message from the dataset where removing it would:
- Reduce the total number of unique **lemmatized tokens**.
- Keep the total number of unique **stemmed tokens** unchanged.

The result highlights cases where stemming and lemmatization handle words differently, emphasizing the linguistic nuances captured by lemmatization but overlooked by stemming.


In [23]:
from comparison import find_message_with_lemma_stem_disparity

In [24]:
# Example usage
result_message = find_message_with_lemma_stem_disparity(
    messages=data["Message"].tolist(), nltk_stems=nltk_stems, nltk_lemmas=nltk_lemmas
)

if result_message:
    print("Message causing the effect:\n")
    print(
        repr(result_message)
    )  # Ensures the entire message, including multiline, is printed
else:
    print("No message found that meets the criteria.")

Message causing the effect:

'Go until jurong point, crazy.. Available only in bugis n great world la e buffet... Cine there got amore wat...'


<br>----------------------------------------------<br>
## Section 10. Comparison of the results NLTK vs. SpaCy

#### Tokenization:
- **NLTK Unique Tokens**: 689
- **SpaCy Unique Tokens**: 743
- **Common Tokens**: 10,822
- **Total Tokens**:
  - NLTK: 104,904
  - SpaCy: 104,367

**Observation**:
- SpaCy identifies slightly more unique tokens due to its advanced handling of special characters, contractions, and linguistic nuances.

#### Lemmatization:
- **NLTK Unique Lemmas**: 2,379
- **SpaCy Unique Lemmas**: 1,434
- **Common Lemmas**: 8,221
- **Total Lemmas**:
  - NLTK: 104,904
  - SpaCy: 104,367

**Observation**:
- SpaCy's lemmatization is more conservative, producing fewer unique lemmas, while NLTK is more aggressive and often generates more variations.

#### Stemming:
- **NLTK Unique Stems**: 5,870
- **SpaCy Unique Stems**: 4,777
- **Common Stems**: 2,197
- **Total Stems**:
  - NLTK: 104,904
  - SpaCy: 104,367

**Observation**:
- NLTK's stemming is more aggressive, leading to more fragmented stems. SpaCy’s stems (approximated using `.lemma_`) are closer to lemmatization and thus less varied.


<br>----------------------------------------------<br>
## Section 11. Comparison of NLTK Before and After Tokenization, Lemmatization, and Stemming

#### Before Tokenization:
- Messages are raw text with varying formats (e.g., punctuation, special characters, etc.).

#### After Tokenization:
- **Total Tokens**: 104,904
- **Unique Tokens**: 689
- NLTK tokenization splits text into words but does not handle linguistic nuances (e.g., contractions are split into separate tokens).

#### After Lemmatization:
- **Total Lemmas**: 104,904
- **Unique Lemmas**: 2,379
- Lemmatization reduces tokens to their dictionary forms but keeps linguistic variations (e.g., "better" → "good").

#### After Stemming:
- **Total Stems**: 104,904
- **Unique Stems**: 5,870
- Stemming is more aggressive, often stripping suffixes and creating non-dictionary words (e.g., "running" → "run", "better" → "better").

**Observation**:
- Tokenization introduces a structured list of words.
- Lemmatization normalizes words to base forms, leading to fewer unique entries.
- Stemming is more aggressive and results in more unique variations than lemmatization.


<br>-----------------------------------------<br>
## section 12. Comparison of SpaCy Before and After Tokenization, Lemmatization, and Stemming

#### Before Tokenization:
- Messages are raw text, similar to the NLTK input.

#### After Tokenization:
- **Total Tokens**: 104,367
- **Unique Tokens**: 743
- SpaCy handles linguistic details better, preserving contractions (e.g., "I'm" remains a single token).

#### After Lemmatization:
- **Total Lemmas**: 104,367
- **Unique Lemmas**: 1,434
- SpaCy produces fewer unique lemmas, as it normalizes words more consistently to their dictionary forms (e.g., "better" → "good", "running" → "run").

#### After Stemming (Approximated):
- **Total Stems**: 104,367
- **Unique Stems**: 4,777
- SpaCy's stemming (approximated using `.lemma_[:n]`) is closer to its lemmatization, resulting in fewer unique variations.

**Observation**:
- Tokenization splits text more precisely, maintaining contractions as single tokens.
- Lemmatization reduces redundancy by normalizing words effectively.
- Stemming (if approximated) produces results closer to lemmatization compared to NLTK.


----------------------------
## Section 13 - 16  BeautifulSoup parsing analysis 

In [25]:
from bs4 import BeautifulSoup
import requests
from tokenize_nltk import tokenize_nltk
from tokenize_spacy import tokenize_spacy
from analysis import compare_tokenizations
from lemmatization_nltk import lemmatize_nltk
from lemmatization_spacy import lemmatize_spacy
from analysis import compare_lemmatizations
from stem_nltk import stem_nltk
from stem_spacy import stem_spacy
from analysis import compare_stemmings
from comparison import find_message_with_stem_lemma_disparity
from comparison import find_message_with_lemma_stem_disparity
import re

In [26]:
# A function to get the published date/time of the article either by tag <time> or from the text itself.
def get_time(soup_cp):
    date_tag = soup_cp.find("time")
    if date_tag:
        date = date_tag.text.strip()
    else:
        paragraphs = soup_cp.find_all("p")
        date = "Date not found"
        for para in paragraphs:
            if "Published:" in para.text:
                date = para.text.split("Published:")[-1].strip()
                break
    print("Publication Date:", date)


# URL of the webpage to scrape
url = "https://dotesports.com/league-of-legends/news/faker-wins-yet-another-award-to-add-to-his-trophy-cabinet"

# Fetch the webpage
response = requests.get(url)

article_text = ""

# Check if the request was successful
if response.status_code == 200:
    # Parse the HTML content
    soup = BeautifulSoup(response.text, "html.parser")

    # Extract the article's title
    title = soup.find("h1").text.strip()
    print("Title:", title)

    get_time(soup)
    paragraphs = soup.find_all("p")

    # Extract all paragraphs in the article
    article_text = "\n".join([para.text.strip() for para in paragraphs])
    print(
        "Article Text (First 594 characters):", article_text[:594]
    )  # I just didn't want the paragraph to be cut in the middle of the sentence.
else:
    print(f"Failed to fetch the page. Status code: {response.status_code}")

if article_text:
    print("\n********** Task 5 **********")
    # Apply Task 5
    # Tokenize using NLTK
    nltk_tokens = tokenize_nltk(article_text)
    print(f"Total NLTK tokens: {len(nltk_tokens)}")

    # Tokenize using spaCy
    spacy_tokens = tokenize_spacy(article_text)
    print(f"Total spaCy tokens: {len(spacy_tokens)}")

    # Compare results.
    compare_token_method_result = compare_tokenizations(nltk_tokens, spacy_tokens)
    for key, value in compare_token_method_result.items():
        print(f"{key}: {value}")

    print("\n********** Task 6 **********")
    # Apply Task 6
    # Lemmatize using NLTK
    nltk_lemmas = lemmatize_nltk(article_text)
    print(f"Total NLTK lemmas: {len(nltk_lemmas)}")

    # Lemmatize using spaCy
    spacy_lemmas = lemmatize_spacy(article_text)
    print(f"Total spaCy lemmas: {len(spacy_lemmas)}")

    # Compare lemmatizations
    lemma_diff = compare_lemmatizations(nltk_lemmas, spacy_lemmas)
    print("\nComparison of Lemmatizations:")
    for key, value in lemma_diff.items():
        print(f"{key}: {value}")

    print("\n********** Task 7 **********")
    # Stem using NLTK (Porter Stemmer)
    nltk_stems = stem_nltk(article_text, method="porter")
    print(f"Total NLTK (Porter) stems: {len(nltk_stems)}")

    # Stem using spaCy (approximation)
    spacy_stems = stem_spacy(article_text)
    print(f"Total spaCy stems: {len(spacy_stems)}")

    # Compare stemmings
    stem_diff = compare_stemmings(nltk_stems, spacy_stems)
    print("\nComparison of Stemmings:")
    for key, value in stem_diff.items():
        print(f"{key}: {value}")

    print("\n********** Task 8 **********")
    messages = article_text.split("\n")  # Splits by newlines
    messages = [line.strip() for line in messages if line.strip()]  # Remove empty lines
    result_message = find_message_with_stem_lemma_disparity(
        messages=messages, nltk_stems=nltk_stems, nltk_lemmas=nltk_lemmas
    )

    if result_message:
        print("Message causing the effect:\n")
        print(
            repr(result_message)
        )  # Ensures the entire message, including multiline, is printed
    else:
        print("No message found that meets the criteria.")

    print("\n********** Task 9 **********")
    # Example usage
    result_message = find_message_with_lemma_stem_disparity(
        messages=messages, nltk_stems=nltk_stems, nltk_lemmas=nltk_lemmas
    )

    if result_message:
        print("Message causing the effect:\n")
        print(
            repr(result_message)
        )  # Ensures the entire message, including multiline, is printed
    else:
        print("No message found that meets the criteria.")

Title: Faker wins yet another award to add to his trophy cabinet
Publication Date: Nov 22, 2024 11:15 am
Article Text (First 594 characters): Published: Nov 22, 2024 11:15 am
Another trophy has been added to the already stacked cabinet of Lee “Faker” Sang-hyeok. The arguably greatest player of all time has been awarded PC Player of the Year for the second year in a row at the Esports Awards.
The recently crowned world champion was voted ahead of the likes of Tyson “TenZ” Ngo, CS2 star Ilya “m0NESY” Osipov, and fellow mid lane rival Jeong “Chovy” Ji-hoon. T1 CEO Joe Marsh was in attendance in Riyadh, Saudi Arabia, to accept the award on Faker’s behalf, where the executive cited the adversity Faker has had to overcome this year.

********** Task 5 **********
Total NLTK tokens: 1520
Total spaCy tokens: 1832
NLTK unique tokens count: 0
spaCy unique tokens count: 2
Common tokens count: 62
Total NLTK tokens: 1520
Total spaCy tokens: 1832

********** Task 6 **********
Total NLTK lemmas: 1520


********** task 10 - 11 *************<br>
Key Observations from the Provided Outputs:
 1. Tokenization:
- **Total NLTK tokens:** 1520  
- **Total spaCy tokens:** 1832  
- **NLTK unique tokens count:** 0
- **spaCy unique tokens count:** 2  
- **Common tokens:** 62  

Observation:
- The difference in token count could be due to how punctuation, contractions, or whitespace is handled.  

---

 2. Lemmatization:
- **Total NLTK lemmas:** 1520  
- **Total spaCy lemmas:** 1832  
- **NLTK unique lemmas count:** 16  
- **spaCy unique lemmas count:** 4  
- **Common lemmas:** 46  

***Observation***:
- The higher number of total lemmas in spaCy suggests that spaCy treats more tokens as distinct.  
- NLTK generates more unique lemmas than spaCy, possibly due to differences in their underlying algorithms or handling of vocabulary.

---

3. Stemming:
- **Total NLTK stems:** 1520  
- **Total spaCy stems:** 1832  
- **NLTK unique stems count:** 4  
- **spaCy unique stems count:** 9  
- **Common stems:** 41  

***Observation***:
- Stemming shows a much smaller set of unique stems compared to lemmas, which is expected as stemming reduces tokens more aggressively.  
- Differences in stem counts and common stems highlight algorithmic variances between NLTK and spaCy.


## Section 17 - 18 Whatsapp Analysis

In [27]:
# Regex pattern to match the WhatsApp message format
whatsapp_pattern = r"^\d{1,2}/\d{1,2}/\d{2}, \d{2}:\d{2} - .*?: (.*)"

# Whatsapp file path
whatsapp_file_path = "data/whatsapp_group_chat.txt"


def extract_messages_from_whatsapp_chat(file_path):
    # Read the WhatsApp chat file
    with open(file_path, "r", encoding="utf-8") as file:
        chat_data = file.readlines()

    chat_messages = []
    for line in chat_data:
        if "<" in line or ">" in line:
            continue
        # Match the pattern for WhatsApp messages
        match = re.match(whatsapp_pattern, line)
        if match:
            chat_messages.append(match.group(1))  # Extract the message content
    return chat_messages


extracted_whatsapp_messages = extract_messages_from_whatsapp_chat(whatsapp_file_path)

if len(extracted_whatsapp_messages) != 0:
    print("\n********** Task 5 **********")
    # Apply Task 5
    # Tokenize using NLTK
    nltk_tokens = tokenize_nltk(extracted_whatsapp_messages)
    print(f"Total NLTK tokens: {len(nltk_tokens)}")

    # Tokenize using spaCy
    spacy_tokens = tokenize_spacy(extracted_whatsapp_messages)
    print(f"Total spaCy tokens: {len(spacy_tokens)}")

    # Compare results.
    compare_token_method_result = compare_tokenizations(nltk_tokens, spacy_tokens)
    for key, value in compare_token_method_result.items():
        print(f"{key}: {value}")

    print("\n********** Task 6 **********")
    # Apply Task 6
    # Lemmatize using NLTK
    nltk_lemmas = lemmatize_nltk(extracted_whatsapp_messages)
    print(f"Total NLTK lemmas: {len(nltk_lemmas)}")

    # Lemmatize using spaCy
    spacy_lemmas = lemmatize_spacy(extracted_whatsapp_messages)
    print(f"Total spaCy lemmas: {len(spacy_lemmas)}")

    # Compare lemmatizations
    lemma_diff = compare_lemmatizations(nltk_lemmas, spacy_lemmas)
    print("\nComparison of Lemmatizations:")
    for key, value in lemma_diff.items():
        print(f"{key}: {value}")

    print("\n********** Task 7 **********")
    # Stem using NLTK (Porter Stemmer)
    nltk_stems = stem_nltk(extracted_whatsapp_messages, method="porter")
    print(f"Total NLTK (Porter) stems: {len(nltk_stems)}")

    # Stem using spaCy (approximation)
    spacy_stems = stem_spacy(extracted_whatsapp_messages)
    print(f"Total spaCy stems: {len(spacy_stems)}")

    # Compare stemmings
    stem_diff = compare_stemmings(nltk_stems, spacy_stems)
    print("\nComparison of Stemmings:")
    for key, value in stem_diff.items():
        print(f"{key}: {value}")

    print("\n********** Task 8 **********")
    result_message = find_message_with_stem_lemma_disparity(
        messages=extracted_whatsapp_messages,
        nltk_stems=nltk_stems,
        nltk_lemmas=nltk_lemmas,
    )

    if result_message:
        print("Message causing the effect:\n")
        print(
            repr(result_message)
        )  # Ensures the entire message, including multiline, is printed
    else:
        print("No message found that meets the criteria.")

    print("\n********** Task 9 **********")
    # Example usage
    result_message = find_message_with_lemma_stem_disparity(
        messages=extracted_whatsapp_messages,
        nltk_stems=nltk_stems,
        nltk_lemmas=nltk_lemmas,
    )

    if result_message:
        print("Message causing the effect:\n")
        print(
            repr(result_message)
        )  # Ensures the entire message, including multiline, is printed
    else:
        print("No message found that meets the criteria.")


********** Task 5 **********
Total NLTK tokens: 65960
Total spaCy tokens: 65762
NLTK unique tokens count: 723
spaCy unique tokens count: 440
Common tokens count: 12443
Total NLTK tokens: 65960
Total spaCy tokens: 65762

********** Task 6 **********
Total NLTK lemmas: 65960
Total spaCy lemmas: 65762

Comparison of Lemmatizations:
NLTK unique lemmas count: 883
spaCy unique lemmas count: 547
Common lemmas count: 12260
Total NLTK lemmas: 65960
Total spaCy lemmas: 65762

********** Task 7 **********
Total NLTK (Porter) stems: 65960
Total spaCy stems: 65762

Comparison of Stemmings:
NLTK unique stems count: 8958
spaCy unique stems count: 5804
Common stems count: 3951
Total NLTK stems: 65960
Total spaCy stems: 65762

********** Task 8 **********
Message causing the effect:

'שלושת האפשרויות הראשונות זה לזרוק כסף לפח'

********** Task 9 **********
Message causing the effect:

'שלושת האפשרויות הראשונות זה לזרוק כסף לפח'


*********** Task 10 - 11 ************

***Analysis of Tokenization, Lemmatization, and Stemming Results***

The provided results show differences in tokenization, lemmatization, and stemming between NLTK and spaCy. Let’s analyze each step:

---
1. Tokenization
- **NLTK Tokens:** 65,960  
- **spaCy Tokens:** 65,762  
- **Unique Tokens:**  
  - NLTK: 723  
  - spaCy: 440  
- **Common Tokens:** 12,443  

Observations:
- **Total Tokens:**  
  The total token counts are nearly identical, suggesting that NLTK and spaCy tokenize the chat text in a similar manner. Minor differences could be due to:  
  - spaCy's handling of punctuation or special characters (e.g., emojis or symbols).  
  - Differences in splitting contractions or compound words.  
- **Unique Tokens:**  
  - NLTK has significantly more unique tokens than spaCy, indicating:  
    - NLTK may preserve more punctuation or symbols as separate tokens.  
    - spaCy may standardize or normalize tokens during processing.  

---

2. Lemmatization
- **NLTK Lemmas:** 65,960  
- **spaCy Lemmas:** 65,762  
- **Unique Lemmas:**  
  - NLTK: 883  
  - spaCy: 547  
- **Common Lemmas:** 12,260  

Observations:
- **Total Lemmas:**  
  Similar counts suggest both libraries are consistent in identifying the base forms of words.  
- **Unique Lemmas:**  
  - NLTK produces more unique lemmas than spaCy, likely due to differences in their lemma dictionaries.  
  - spaCy may use stricter normalization, merging certain word forms that NLTK keeps distinct.  
- **Common Lemmas:**  
  Overlap (12,260) shows both libraries share many base forms but diverge in handling specific tokens.  

---

3. Stemming
- **NLTK Stems:** 65,960  
- **spaCy Stems:** 65,762  
- **Unique Stems:**  
  - NLTK: 8,958  
  - spaCy: 5,804  
- **Common Stems:** 3,951  

Observations:
- **Total Stems:**  
  Like tokens and lemmas, total counts are nearly identical, reflecting similar word segmentation.  
- **Unique Stems:**  
  - NLTK’s Porter Stemmer generates more unique stems, likely due to its aggressive stemming (e.g., cutting suffixes more aggressively).  
  - spaCy stems fewer unique forms, as it uses simpler techniques for stemming.  
- **Common Stems:**  
  The overlap (3,951) suggests that while both approaches share some base forms, they diverge significantly in how they reduce words to their stems.
